docker/github/__init__.py

```diff
@@ -114,25 +113,36 @@ async def get_recent_deployment(self, environment: Optional[str]) -> Optional[De
         return deployments[0] if deployments else None
 
     async def verify_ref_is_deployed_in_previous_environment(
-        self, ref: str, environment: str, ordered_environments: Sequence[str],
+        self,
+        ref: str,
+        environment: str,
+        ordered_environments: Sequence[str],
     ):
         index = ordered_environments.index(environment)
         if index == 0:
             return
 
         previous_environment = ordered_environments[index - 1]
-        if not await self.is_deployed_in_environment(ref, previous_environment):
+        if not await self.is_successfully_deployed_in_environment(ref, previous_environment):
             raise ConstraintError(
                 f"Deployment of {ref} to {environment} failed, because deployment to {previous_environment} is missing",
             )
 
-    async def is_deployed_in_environment(self, ref: str, environment: str) -> bool:
-        return any(ref == deployment.ref for deployment in await self.get_deployments(environment))
+    async def is_successfully_deployed_in_environment(self, ref: str, environment: str) -> bool:
+        for deployment in await self.get_deployments(environment):
+            if (
+                ref == deployment.ref
+                and (statuses := await self.get_deployment_statuses(deployment.id))
+                and statuses[0].state == DeploymentState.success
```

Comment:
Was ist, wenn es hier noch gar keinen Status gibt, weil das Deployment erst gerade gestartet wurde? Dann fällt der Zugriff auf `[0]` auf die Nase.
Wäre es nicht sinnvoll hier die Sortierung gar nicht zu beachten, sondern einfach mit `any` zu checken, ob es mindestens einen Success-Status gibt? Ich würde davon ausgehen, dass ein einmal gesetzer Success nie mehr überschrieben werden kann. Oder ist diese Annahme nicht korrekt?

Statt `for, if, return True, return False` würde ich ebenfalls `any` nutzen.

Reply:
* > Was ist, wenn es hier noch gar keinen Status gibt, weil das Deployment erst gerade gestartet wurde? Dann fällt der Zugriff auf [0] auf die Nase. Das war Bestandteil meiner Tests.
  
  Nein, das ist durch `statuses := ...` abgefangen.
* > Ich würde davon ausgehen, dass ein einmal gesetzer Success nie mehr überschrieben werden kann.
  
  Es können beliebig viele Statuse in beliebiger Reihenfolge angelegt werden. Bis auf die Struktur des Payloads wird seitens GitHub nichts geprüft oder eingeschränkt was das angeht.
* `any` hat nicht funktioniert, durch das `if` hat sich Python über `'async_generator' object is not iterable` beschwert, das habe ich leider nicht anders lösen können.

Reply:
> Nein, das ist durch statuses := ... abgefangen.

Ahh, das habe ich übersehen, Danke für den Hinweis.

> any hat nicht funktioniert, durch das if hat sich Python über 'async_generator' object is not iterable beschwert, das habe ich leider nicht anders lösen können.

# Idee 1

Ich habe mir das ein bißchen anhand dieses vereinfachten Beispiels angesehen:
```python
import asyncio

async def items():
    return (1,)

async def main():
    print(any((
        deployment
        for deployment in await items()
        if (statuses := await items())
        and statuses[0] == 1
    )))

asyncio.run(main())

>>> TypeError: 'async_generator' object is not iterable
```

Damit kann ich den Fehler reproduzieren. Wenn man jetzt den Generator für `any` stattdessen als List-Comprehension schreibt, funktioniert das verrückterweise:

```python
import asyncio

async def items():
    return (1,)

async def main():
    print(any([
        deployment
        for deployment in await items()
        if (statuses := await items())
        and statuses[0] == 1
    ]))

asyncio.run(main())

>>> True
```

Scheinbar bin ich nicht der einzige, der das überraschend findet: 
- https://stackoverflow.com/questions/52285419/aggregating-an-async-generator-to-a-tuple
- https://stackoverflow.com/questions/42448664/async-generator-is-not-an-iterator

Die Erklärung von dem Typen der asynchrone Generatoren und Comprehensions in Python 3.6 eingebaut hat, findet sich hier:
- https://bugs.python.org/issue32113
 
Auch wenn die Syntax für asynchrone Generators und Comprehensions sehr ähnlich ist (runde Klammern statt eckige), kann man anhand der beiden Beispiele gut sehen, dass unter der Haube doch ganz andere Dinge passieren. 

# Idee 2

Hatte auch mal nach einem asynchronen `any()` gesucht und bin dabei auf das hier gestoßen:
- https://asyncstdlib.readthedocs.io/en/latest/source/api/builtins.html

Keine Ahnung, ob das was taugt. Generell würde dieser Ansatz es halt ermöglichen, nicht das synchrone `any()` nutzen zu müssen, wodurch dann eine `async -> sync` Konvertierung in der Comprehension entfällt.

# Idee 3

Grundsätzlich ist der Ansatz im Code hier mit
```python
for deployment in await get_deployments():
    statuses = await get_statuses()
    ...
```
dem Async-Gedanken nicht so ganz treu, da für jeden Status-Request einzeln gewartet wird. Eigentlich könnte man die Deployments abholen und dann gleichzeitig für alle Deployments die Statusse.
Das sähe dann z.B. so aus:

```python
import asyncio
from dataclasses import dataclass

@dataclass
class Deployment:
    ref: int

async def get_deployments() -> tuple[Deployment, ...]:
    return tuple(map(Deployment, range(3)))

async def get_deployment_statuses(deployment: Deployment) -> tuple[int, ...]:
    return (4, 5, 6)

async def main():
    print(any((
        statuses and statuses[0] == 4
        for statuses in await asyncio.gather(*(
            get_deployment_statuses(deployment)
            for deployment in await get_deployments()
            if deployment.ref == 2
        ))
    )))

asyncio.run(main())
```
Das ist ein bißchen mehr in Richtung Parallel-Denke.

# Idee 4
Theoretisch könnte man auch auch überlegen, ob es etwas bringen würde asynchrone Iteratoren wie in PEP-525 zu nutzen ("asynchronous data-producer"). Allerdings müßte man dann auch an die Implementierungen von `get_deployments` und `get_deployment_statuses` ran. Noch mehr Async-Zeug macht das Ganze dann sicherlich auch nicht einfacher zu nutzen.

# Fazit

War interessant für mich, nochmal etwas tiefer in das Thema einzusteigen, habe versucht meine Erkenntnisse ein bißchen zu teilen.
Wie ist dein Blick auf das Thema? Ist da was dabei, wo du sagen würdest, dass es die Sache besser machen könnte?

Reply:
Ich habe blöderweise keine Zeit gehabt mir inhaltlich hier Gedanken zu machen, aber ist `gather` eigentlich nicht immer der Tool der Wahl statt Generatoren oder Comprehensions wenn man irgendwas auf die `async` mengen tuen will? Insoweit für dieses konkretes Beispiel dachte ich `any(gather())` ist definitiv die "richtige" Lösung? Bin ein bisschen verwirrt durch die Vielhalt deiner Ideen...

Reply:
Ja, ich persönlich würde auch sagen, dass Idee 3 mit `gather` der "richtige" (=parallele) Ansatz ist. Wenn man überall einfach immer `await` nutzt, wenn irgendwo ein `awaitable` rauskommt, kann man sich den komplizierten Async-Kram auch sparen und an den wenigen kritischen Stellen dann Thread-Pools/Executors unter der Haube verstecken. Die Performance ist dann identisch und die Komplexität aufgrund synchroner Funktionen/Interfaces um den Faktor 5 geringer.
Wenn man Idee 3 folgt, wird eigentlich niemals sowas wie `for ... if await` irgendwo auftauchen, weil das einfach per Definition nicht richtig parallel ist.

Idee 2 ist grundsätzlich sogar mit Idee 3 kombinierbar, da noch paralleler. Sobald *ein* passendes Ergebnis da ist, müssen die anderen Futures alle nicht mehr ausgeführt werden bzw. könnten sogar gecancelled werden. Die referenzierte Bibliothek geht aber leider auch nicht so weit, dass sie Dinge wieder cancelled. So ernst nimmt dann halt scheinbar doch niemand eine vollständige Parallelisierung...

Wenn man nicht so grundsätzlich militant zum Thema Parallelität steht wie ich, dann macht Idee 1 insofern Sinn, als dass der Code mit einer Zwei-Zeichen-Änderung funktionieren würde. Hat natürlich die gleichen Probleme wie eben beschrieben und vermeidet nur das direkte Problem mit dem Fehler `TypeError: 'async_generator' object is not iterable`. 

Idee 4 ist nicht wirklich ausgearbeitet, weil ich nicht weiter Zeit investieren wollte. Ich glaube auch im Moment nicht, dass da was tolles rauskommen würde.

Reply:
Alles klar, ich durch meine FZJ-Background bin in Sachen `async` Leidenschaftslos und mache einfach stumpfe OpenMP-mäßige `concurrent.futures` map, da wo man es braucht. Aber wie gesagt, wenn man schon alles `async` macht, dann nehme ich `gather`...

Fazit: verstanden, wir sind uns einig, du wolltest nur mit die Gedanken rumspielen.